p2p/transport.go

```diff
@@ -0,0 +1,579 @@
+	// peer currently. All relevant configuration should be refactored into options
+	// with sane defaults.
+	mConfig   conn.MConnConfig
+	p2pConfig config.P2PConfig
+}
+
+// Test multiplexTransport for interface completeness.
+var _ Transport = (*MultiplexTransport)(nil)
+var _ transportLifecycle = (*MultiplexTransport)(nil)
+
+// NewMultiplexTransport returns a tcp conencted multiplexed peers.
+func NewMultiplexTransport(
+	nodeInfo NodeInfo,
+	nodeKey NodeKey,
+) *MultiplexTransport {
+	return &MultiplexTransport{
+		acceptc:          make(chan accept),
+		closec:           make(chan struct{}),
+		dialTimeout:      defaultDialTimeout,
+		filterTimeout:    defaultFilterTimeout,
+		handshakeTimeout: defaultHandshakeTimeout,
+		mConfig:          conn.DefaultMConnConfig(),
+		nodeInfo:         nodeInfo,
+		nodeKey:          nodeKey,
+		conns:            map[string]net.Conn{},
+		peers:            map[ID]Peer{},
+	}
+}
+
+// Accept implements Transport.
+func (mt *MultiplexTransport) Accept(cfg peerConfig) (Peer, error) {
+	select {
+	// This case should never have any side-effectful/blocking operations to
+	// ensure that quality peers are ready to be used.
+	case a := <-mt.acceptc:
+		if a.err != nil {
+			return nil, a.err
+		}
+
+		cfg.outbound = false
+
+		p := mt.wrapPeer(a.conn, a.nodeInfo, cfg)
+
+		return p, mt.filterPeer(p)
+	case <-mt.closec:
+		return nil, &ErrTransportClosed{}
+	}
+}
+
+// Dial implements Transport.
+func (mt *MultiplexTransport) Dial(
+	addr NetAddress,
+	cfg peerConfig,
+) (Peer, error) {
+	c, err := addr.DialTimeout(mt.dialTimeout)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO(xla): Evaluate if we should apply filters if we explicitly dial.
+	if err := mt.filterConn(c); err != nil {
+		return nil, err
+	}
+
+	sc, ni, err := mt.upgrade(c)
+	if err != nil {
+		return nil, err
+	}
+
+	cfg.outbound = true
+
+	p := mt.wrapPeer(sc, ni, cfg)
+
+	return p, mt.filterPeer(p)
+}
+
+// Close implements transportLifecycle.
+func (mt *MultiplexTransport) Close() error {
+	close(mt.closec)
+
+	return mt.listener.Close()
+}
+
+// Listen implements transportLifecycle.
+func (mt *MultiplexTransport) Listen(addr NetAddress) error {
+	ln, err := net.Listen("tcp", addr.DialString())
+	if err != nil {
+		return err
+	}
+
+	mt.listener = ln
+	mt.nodeAddr = addr
+
+	go mt.acceptPeers()
+
+	return nil
+}
+
+func (mt *MultiplexTransport) acceptPeers() {
+	for {
+		c, err := mt.listener.Accept()
+		if err != nil {
+			// Close has been called so we silently exit.
+			if _, ok := <-mt.closec; !ok {
+				return
+			}
+
+			mt.acceptc <- accept{err: err}
+			return
+		}
+
+		// Connection upgrade and filtering should be asynchronous to avoid
+		// Head-of-line blocking[0].
+		// Reference:  https://github.com/tendermint/tendermint/issues/2047
+		//
+		// [0] https://en.wikipedia.org/wiki/Head-of-line_blocking
+		go func(c net.Conn) {
+			var (
+				ni NodeInfo
+				sc *conn.SecretConnection
+			)
+
+			err := mt.filterConn(c)
```

Comment:
Does filterConn need to be async? These should be quick local checks (granted could be over ABCI, but that should be quick too).

Also, can/should a conn filter closure do the maxPeers check (currently happening in the switch)?

Reply:
> Does filterConn need to be part of this async routine? These should be quick local checks (granted could be over ABCI, but that should be quick too).

As the filters are handed from the outside and potentially are side-effectful (network) I advise to be as defensive as possible, as in executing them asynchronously and guard against long running operations.

> Also, can/should a conn filter closure do the maxPeers check (currently happening in the switch)?

Very valid point, the more the transport shapes up the more of the peer regulation seems to fit there. But currently the switch has still a lot of logic regarding peer managament, so I'd leave there for now.